packages/sakuli-legacy/src/context/sahi/action/utils/create-with-action-context.function.ts

```diff
@@ -1,31 +1,39 @@
 import { TestExecutionContext } from "@sakuli/core";
 import { ifPresent } from "@sakuli/commons";
+import * as util from "util";
 
 export function createWithActionContext(ctx: TestExecutionContext) {
   return function runAsAction<ARGS extends any[], R>(
     name: string,
     fn: (...args: ARGS) => Promise<R>
   ): (...args: ARGS) => Promise<R> {
+    function logFinishedAction(error: any) {
+      const log = [`Finish action ${name}`];
+      ifPresent(ctx.getCurrentTestAction(), (action) => {
+        log.push(error ? `with errors` : `successfully`);
+        log.push(
+          `after ${
+            (new Date().getTime() - action.startDate!.getTime()) / 1000
+          }s`
+        );
+      });
+
+      util.types.isNativeError(error)
+        ? ctx.logger.trace(log.join(" "), error.stack)
+        : ctx.logger.trace(log.join(" "));
+    }
```

Comment:
This function covers two different responsibilities. One responsibility is logging a successful test action, another is logging failed test actions. This comes with two `?` operators distinguishing both responsibilities again so that they can be covered in one method.
This adds a lot of complexity to the `logFinishedAction` method and it's surroundings, because you have an extra `actionError` variable which is undefined as long as an error occurs. In `logFinishedAction` you check whether an error occurred by checking the `actionError`, which is only an error if the catch blog was hit. So what the `actionError` variable does for you is to memorize that the code flow hit the catch blog.

I'd suggest to restructure the code to separate the responsibilities of `logFinishedAction`. 
```typescript
try {
 ...
 logSuccess()
} catch(e){
 logError(e)
}finally{
  ctx.endTestAction();
}
```

There is, nevertheless, some shared code in that function which could be extracted into a method to be reused or just copied. Shared code is line 11 and 14-18. I'd copy line 11 and extract the calculation of the step duration into a function.

Just to point out the complexity that is compressed in `logFinishedAction`:
The method is 14 lines long. 6 lines is shared code, while 5 out of that 6 lines are mostly braces and formatting by prettier. So not even 50% of the code (even with formatting) is common code between the both responsibilities of logging errors and logging success. The other lines perform or require a dispatching of the different responsibilities before they can take action.